### DISCUSSION TITLE
EASTL containers implementation

### DISCUSSION THREAD

**Discussion by cupsster** [nlohmann/json#2751] [Q&A]
Hi, I would like to ask if there is any simple way of how to make [EASTL](https://github.com/electronicarts/EASTL/blob/master/doc/FAQ.md) containers work with json? Most of them mirror stl implementation but I still get errors when defining non intrusive macros for my types that use EASTL containers as members. If some simple example can be provided how to define conversions for basic container like `eastl::vector<T>` that could keep me going for the rest. json read.me mentions how to serialize iterative containers but no word or example how to deserialize. or if anyone went through this and have header with  **EASTL** conversions I'd gladly fork that.

**Comment by nlohmann** [comment]
For templated types such as `eastl::vector<T>`, you need to define a custom `adl_serializer` class, see https://github.com/nlohmann/json#how-do-i-convert-third-party-types.

**Reply by cupsster** [reply]
I did this but I'm absolutely not sure if this is how it should be done:
``` cpp
#pragma once

#include <EASTL/vector.h>
#include <nlohmann/json.hpp>

// partial specialization (full specialization works too)
namespace nlohmann
{
    template <typename T>
    struct adl_serializer<eastl::vector<T>>
    {
        static void to_json( json& j, const eastl::vector<T>& data )
        {
            if (!data.empty())
                j = json::parse(data);
        }

        static void from_json( const json& j, eastl::vector<T>& data )
        {
            if (!j.is_null())
            {
                data = j.get<eastl::vector<T>>();
            }
        }
    };
}
```

**Reply by cupsster** [reply]
Used `json::parse` for serialization as container has standard iterator but have no clue what should be done to de-serialize it properly.

**Reply by nlohmann** [reply]
- The `to_json` function seems to make the assumption that the vector contains some bytes to parse from. What would be the type `T` there?
- The `from_json` function looks recursive - you define the JSON-to-vector function by calling `get`.

I would rather expect the following:

```cpp
// partial specialization (full specialization works too)
namespace nlohmann
{
    template <typename T>
    struct adl_serializer<eastl::vector<T>>
    {
        static void to_json( json& j, const eastl::vector<T>& data )
        {
        	// create a JSON array from the elements
        	for (const auto& el : data)
        	{
        		j.push_back(el); // assuming there already is a conversion from T to JSON
        	}
        }

        static void from_json( const json& j, eastl::vector<T>& data )
        {
        	for (const auto& el : j)
        	{
        		data.push_back(el.get<T>()); // assuming there already is a conversion from JSON to T
        	}
        }
    };
}
```

This should (did not test it) make `eastl::vector<T>` behave just like `std::vector<T>`.
